### ISSUE TITLE
`find` operations are not timing out when buffered

### ISSUE THREAD

**Issue by AndrewThorwall** [Automattic/mongoose#14184]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

6.12.0

### Node.js version

18.18.2

### MongoDB server version

6.0.12

### Typescript version (if applicable)

5.3.3

### Description

I'm seeing that `find` operations do not seem to time out when they're buffered for longer than `bufferTimeoutMS`.

- If I set `bufferCommands: false`, and then attempt to invoke a `find` operation before connecting to Mongo, I get an exception (expected). 
- If I set `bufferCommands: true` and `bufferTimeoutMS: 1000`, and then attempt to invoke a `find` operation before connecting to Mongo, the operation hangs indefinitely (unexpected).
- If I set `bufferCommands: true` and `bufferTimeoutMS: 1000`, and then attempt to invoke an `update` or `delete` operation before connecting to Mongo, the operation times out after one second (expected).

Of course this only matters if we are attempting to perform operations before we connect to Mongo, which we can avoid doing. But this is coming up in the context of helping developers see their issue quickly by failing fast if they have something misconfigured, rather than hanging indefinitely

### Steps to Reproduce

- Create a Mongoose model with a schema that has `bufferCommands: true` and `bufferTimeoutMS: 1000`.
- Attempt to execute a `find` operation without opening a connection to Mongo.

### Expected Behavior

- The invocation throws an exception ~1 seconds after it begins.

**Comment by IslandRhythms** [comment]
On 6.12 it ends the script without an error being thrown and on mongoose 8 it does throw the exception. Can't get it so hang indefinitely.

```javascript
const mongoose = require('mongoose');

const testSchema = new mongoose.Schema({
  name: String
});

const Test = mongoose.model('Test', testSchema);

async function run() {
  mongoose.set('bufferCommands', true);
  mongoose.set('bufferTimeoutMS', 1000);
  await Test.find();
  await mongoose.connect('mongodb://localhost:27017');
  await mongoose.connection.dropDatabase();
  console.log('done');
}

run();
```

**Comment by AndrewThorwall** [comment]
Ah, I misunderstood what is happening under the hood and it is true that execution is not actually "hanging" per se. But instead that execution path terminates silently. In my application that appeared the same as hanging because I have other asynchronous execution happening which causes Node to continue running. 

So, that's the behavior that seems wrong to me. If `find()` fails somewhere due to an exception, I expect that exception to be propagated up, and the returned promise should be rejected. Concretely, if I run this code:

```
const mongoose = require("mongoose");

const testSchema = new mongoose.Schema({
  name: String,
});

const Test = mongoose.model("Test", testSchema);

async function readFromDatabase() {
  mongoose.set("bufferCommands", true);
  mongoose.set("bufferTimeoutMS", 1000);
  try {
    await Test.find();
  } catch (e) {
    console.error("Failure", e);
    throw e;
  }
  console.log("Success");
  await mongoose.connect("mongodb://localhost:27017");
  await mongoose.connection.dropDatabase();
}

readFromDatabase().finally(() => {
  console.log("This will be printed no matter what");
});
```

I expect two things to be true:

- `This will be printed no matter what.` will be printed.
- Either `Success` or `Failure` will be printed. In this case, because we're not connected to the database within the timeout, `Failure` should be printed.

Does this make sense to you? Or am I misunderstanding? For what it's worth, based on your comment, it seems like this is likely fixed in Mongoose 8, which would be good news because we are planning to upgrade soon.

**Comment by vkarpov15** [comment]
@AndrewThorwall I'm not sure what you mean by "But instead that execution path terminates silently.". When running your script against Mongoose 6.12.5, I get the following output:

```
$ node gh-14184.js 
Failure MongooseError: Operation `tests.find()` buffering timed out after 1000ms
    at Timeout.<anonymous> (/home/val/Workspace/MongoDB/troubleshoot-mongoose/node_modules/mongoose/lib/drivers/node-mongodb-native/collection.js:186:23)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7)
This will be printed no matter what
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

MongooseError: Operation `tests.find()` buffering timed out after 1000ms
    at Timeout.<anonymous> (/home/val/Workspace/MongoDB/troubleshoot-mongoose/node_modules/mongoose/lib/drivers/node-mongodb-native/collection.js:186:23)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7)

Node.js v18.17.1

```

So both of your expectations appear to be true. `This will be printed no matter what. ` is printed, as is `Failure`. As far as I can tell Mongoose is correctly throwing a timeout error when buffering times out here - am I missing something?

**Comment by AndrewThorwall** [comment]
I'm not sure what's happening then. For me the program terminates after one second with no output. If I set `bufferCommands` to `false` then I get the expected output:

```
Failure MongooseError: Cannot call `tests.find()` before initial connection is complete if `bufferCommands = false`. Make sure you `await mongoose.connect()` if you have `bufferCommands = false`.
    at [removed stack]

MongooseError: Cannot call `tests.find()` before initial connection is complete if `bufferCommands = false`. Make sure you `await mongoose.connect()` if you have `bufferCommands = false`.
    at [removed stack]

Node.js v20.10.0
```

(I'm running with Node v20.10.0 at the moment but this is the same behavior I was seeing on Node 18)

But when I set it back to `true` the program ends after one second with no output.

For what it's worth it seems like @IslandRhythms got the same behavior as me based on their comment so that points to there being some legitimate issue here and not just a problem with my local environment. Maybe there is a difference between 6.12.0 and 6.12.5 in this regard (I am using Mongoose 6.12.0 at the moment) - that would be great. At the moment I don't have any time to investigate this any further but I will set aside some time later to investigate whether a Mongoose upgrade fixes it.
